feature/#106-Support-Sqlalchemy-1.4-part-1

Nicoretti opened the pull request:
- Update sqlalchemy dependency to base version to 1.4.0
- Update ssl/tls test to comply with new sqla 1.4 api
- Remove backwards compatibility with SQLA < 1.4 for ssl/tls test
- Add changelog entry
- Disable statement cache for pyodbc and turbodbc dialects
- Upgrade tests to use new introspection api
- Update URL creation to new api
- Fix pre_exec edge case
- Restructure test setup & execution

Nicoretti committed:
Update sqlalchemy dependency to base version to 1.4.0

Nicoretti committed:
Update ssl/tls test to comply with new sqla 1.4 api

Nicoretti committed:
Remove backwards compatibility with SQLA < 1.4 for ssl/tls test

Nicoretti committed:
Add changelog entry

Nicoretti committed:
Disable statement cache for pyodbc and turbodbc dialects

For more details see also:
* https://github.com/exasol/sqlalchemy-exasol/issues/190
* https://docs.sqlalchemy.org/en/14/core/connections.html#engine-thirdparty-caching
* https://docs.sqlalchemy.org/en/14/faq/performance.html#why-is-my-application-slow-after-upgrading-to-1-4-and-or-2-x

Nicoretti committed:
Upgrade tests to use new introspection api

Nicoretti committed:
Update URL creation to new api

Nicoretti committed:
Fix pre_exec edge case

In some casese e.g. when the server version haven't been reflected yet
pre_exec also needs to be skipped

tkilias commented on sqlalchemy_exasol/base.py:
```diff
@@ -762,48 +762,50 @@ def pre_exec(self):
-            self.parameters = [[]]
+        server_version = self.root_connection.dialect.server_version_info
+        skip_pre_exec = (
+            not self.isdelete and server_version is None or server_version >= (4, 1, 8)
```
I think, it is save to drop support for Exasol version 4

Nicoretti replied:
nice, still we probably still need it for the delete case, at least that needs to be clarified.

tkilias replied:
yep, think the same

tkilias commented on changelog.d/20220810_120055_nicola.coretti__106_Support_Sqlalchemy_1_4.rst:
```diff
@@ -0,0 +1,10 @@
+
+    .. warning::
+
+        This may impact the performance (see also `SQLAlchemy docs <https://docs.sqlalchemy.org/en/14/faq/performance.html#why-is-my-application-slow-after-upgrading-to-1-4-and-or-2-x>`_).
```
If, that means probably quite a lot work

tkilias replied:
Do you know, if it at least caches the metadata queries?

Nicoretti replied:
I don't no yet. Because the caching topic seems to be quite complex I moved it into an follow up issue, so we can focus on the migration to 1.4 first.

tkilias commented on pyproject.toml:
```diff
@@ -54,13 +54,14 @@ exclude = []
 #  - https://github.com/blue-yonder/turbodbc/issues/358
 #  - https://github.com/exasol/sqlalchemy-exasol/issues/146
 # turbodbc = { version = ">4.5.3", optional = true }
+turbodbc = {path = "turbodbc-4.5.3.tar.gz"}
```
This needs to go later

Nicoretti replied:
Yes, good reminder. But need it for testing right now (at least until https://github.com/blue-yonder/turbodbc/pull/359 gets merged and released).

tkilias commented on sqlalchemy_exasol/base.py:
```diff
@@ -762,48 +762,50 @@ def pre_exec(self):
-                        raise TypeError("Data type not supported: %s" % type(value))
-            self.statement = db_query
-            self.parameters = [[]]
+        server_version = self.root_connection.dialect.server_version_info
```
Did you change anything here? Or, is this reformating? Because the diff is unfortunately not really helpful

Nicoretti replied:
no I just added a extra variable to store the `server_version` so the code using it will be more concise.

Nicoretti committed:
Remove unnecessary code

due to the fact that the result of the calculation result of the else branch is not
used it can be removed with changing the actual behaviour.

Nicoretti committed:
Simplify initialization of on columns

Nicoretti committed:
Simplify column selection in merge class

Nicoretti committed:
Refactor _append_where method

* Make it static
* Simplify code

Nicoretti committed:
Add an additional indirection for the visitor

* Make sure visit will not access private function(s) of merge

Nicoretti committed:
Refactor visit function

* Split up into smaller subfunctions

tkilias commented on sqlalchemy_exasol/merge.py:
```diff
@@ -11,115 +11,136 @@
-                source_cols[e.name] = e
-        return source_cols
+    def _source_columns(self):
+        elements = list(
```
why do you use list + iter comprehension instead by a list comprehension

tkilias commented on sqlalchemy_exasol/merge.py:
```diff
@@ -11,115 +11,136 @@
+                for column in self._target_table.c
+                if column not in self._on_columns and column.name in src_columns
+            )
+            [values.update({column: src_columns[column.name]}) for column in columns]
```
you probably could create values in on dict comprehension with two for's, but this is not always better to read, but you can likely use a dict comprehension instead of values update.

tkilias commented on sqlalchemy_exasol/merge.py:
```diff
@@ -11,115 +11,136 @@
-                    self._on_columns.append(e)
-            else:
-                elements_to_check.extend(e.get_children())
+        self._on_columns = list(
```
I think, the logic doesn't equate the previous version, becaue the previous version looks like a serialized recursive descent into a tree

tkilias commented on sqlalchemy_exasol/merge.py:
```diff
@@ -11,115 +11,136 @@
+            columns = (
+                column for column in self._target_table.c if column.name in src_columns
+            )
+            [values.update({column: src_columns[column.name]}) for column in columns]
```
you probably could create values in on dict comprehension with two for's, but this is not always better to read, but you can likely use a dict comprehension instead of values update.

tkilias commented on sqlalchemy_exasol/merge.py:
```diff
@@ -11,115 +11,136 @@
+            sql += self._compile_delete(compiler)
+        if self._is_merge_insert():
+            sql += self._compile_insert(compiler)
+        return sql
```
I guess, you simply extracted the parts into methods, this way better understandable

Nicoretti committed:
Rename member _target_table to table to leverage sqla internals

* E.g. provide sql rendering preview etc

Nicoretti committed:
Create extra class for delete clause within merge

Nicoretti committed:
Fix translate map regression test

Nicoretti committed:
Add updated poetry.lock file

Nicoretti committed:
Disable unsupported features regarding binary types

Nicoretti committed:
Fixed copy and paste error in github action

Nicoretti committed:
Fix and simplify skipping of update tests for turbodbc

Nicoretti committed:
Add db connection setting(s) for pyodbc and turboodbc tests

Nicoretti committed:
Fix typo in turbodbc connection string

Nicoretti committed:
Add review comments which to be addressed once tests will pass

Nicoretti committed:
Restructure test(s)

Nicoretti committed:
Restructure test setup

Nicoretti committed:
Fix typing compatibility issue

Nicoretti committed:
Update github actions

Nicoretti committed:
Update integration test docs to reflect latest changes and findings

Nicoretti committed:
Update dependencies

Nicoretti committed:
Update dependencies

Nicoretti committed:
Update mypy config

Nicoretti committed:
Fix mypy warning in typeshed

Nicoretti committed:
Remove pyodbc from type checking

Nicoretti committed:
Upgrade pyupgrade dependency

tkilias commented on .github/workflows/ci-cd.yml:
```diff
@@ -49,79 +82,74 @@ jobs:
+      - name: Install poetry
+        uses: abatilo/actions-poetry@v2.1.5
+        with:
+          poetry-version: 1.1.13
```
Poetry 1.2.0

tkilias commented on .github/workflows/ci-cd.yml:
```diff
@@ -49,79 +82,74 @@ jobs:
+      - name: Install poetry
+        uses: abatilo/actions-poetry@v2.1.5
+        with:
+          poetry-version: 1.1.13
```
Poetry 1.2.0

tkilias commented on .github/workflows/ci.yaml:
```diff
@@ -36,9 +36,42 @@ jobs:
+      - name: Install poetry
+        uses: abatilo/actions-poetry@v2.1.5
+        with:
+          poetry-version: 1.1.13
```
Poetry 1.2.0

tkilias commented on .github/workflows/ci.yaml:
```diff
@@ -53,72 +86,71 @@ jobs:
+      - name: Install poetry
+        uses: abatilo/actions-poetry@v2.1.5
+        with:
+          poetry-version: 1.1.13
```
Poetry 1.2.0

tkilias commented on .github/workflows/ci.yaml:
```diff
@@ -53,72 +86,71 @@ jobs:
+        working-directory: ..
+        env:
+          ITDE_URL: "https://github.com/exasol/integration-test-docker-environment.git"
+          ITDE_TAG: "0.11.0"
```
Use ITDE 1.1.0

Nicoretti replied:
I rather would switch versions of ITDE at the end (once all tests are green) to reduce the amount of moving parts regarding testing.

tkilias replied:
ok, fine by me, maybe create a ticket as reminder

tkilias commented on .github/workflows/ci-cd.yml:
```diff
@@ -49,79 +82,74 @@ jobs:
+        working-directory: ..
+        env:
+          ITDE_URL: "https://github.com/exasol/integration-test-docker-environment.git"
+          ITDE_TAG: "0.11.0"
```
Use 1.1.0

Nicoretti replied:
Due to the fact that this potentially impacts testing itself I rather would pros pone this until all test are green again.

tkilias commented on doc/developer_guide/integration_tests.rst:
```diff
@@ -1,5 +1,41 @@
+The integration tests are located within `test/integration`. They are split in
+two groups.
+
+#. The sqlachmey conformance test suite
```
SQLAlchemy is a named entity and probably isn't written lower case

tkilias commented on doc/developer_guide/integration_tests.rst:
```diff
@@ -1,5 +1,41 @@
+    The sqlalchemy conformance test suite is provided and maintained by the sqlalchemy project and intended to support third party dialect developers.
+    For further details see also `README.dialects.rst <https://docs.exasol.com/db/latest/sql_reference.htm>`_.
+
+#. Our custom exasol test suite
```
Exasol is a named entity and probably isn't written lower case

tkilias commented on doc/developer_guide/integration_tests.rst:
```diff
@@ -1,5 +1,41 @@
+.. attention::
+
+    The exasol database does do implicit schema/context changes (open & close schema)
+    in certain scenarios. Keep this in mind when setting up (writing) tests.
```
Using a new connection can avoid aide effects.

tkilias commented on doc/developer_guide/integration_tests.rst:
```diff
@@ -1,5 +1,41 @@
+    For further details have a look at the `Exasol-Documentation <https://docs.exasol.com/db/latest/sql_reference.htm>`_.
+
+
+Setup
```
With setup, I had thought you mean the local test setup and not only the GitHub actions

tkilias commented on doc/developer_guide/integration_tests.rst:
```diff
@@ -1,5 +1,41 @@
+Setup
+=====
 
 Integration testing is done by GitHub Actions contained in this repository, which provide a CI/CD pipeline to test, build, and deploy sqlalchemy_exasol to Pypi.
```
PyPI is also a named entity and as this it should written as it is on their website.

tkilias commented on noxfile.py:
```diff
@@ -230,20 +115,26 @@ def is_version_in_sync() -> bool:
     session.notify("code-format")
     session.notify("type-check")
     session.notify("lint")
-    session.notify("type-check")
```
Did you remove mypy. Is there a specific reason?

tkilias commented on noxfile.py:
```diff
@@ -287,12 +179,13 @@ def populate() -> None:
                 (
                     "CREATE SCHEMA TEST_SCHEMA;",
-                    "CREATE SCHEMA TEST_SCHEMA_2;",
+                    # "CREATE SCHEMA TEST_SCHEMA_2;",
```
If we don't need this, then remove it.

tkilias commented on noxfile.py:
```diff
@@ -303,89 +196,148 @@ def stop_db(session: Session) -> None:
-            ]
-        ).format(connector=connector, db_port=Settings.DB_PORT)
-        session.run("pytest", "--dropfirst", "--dburi", uri, external=True, env=env)
+    Attention:
```
We should later maybe codify this condition

tkilias replied:
ah, I think this one is outdated, I saw you call them seperately

tkilias commented on .github/workflows/ci.yaml:
```diff
@@ -53,72 +86,71 @@ jobs:
+      - name: Install poetry
+        uses: abatilo/actions-poetry@v2.1.5
+        with:
+          poetry-version: 1.1.13
```
Poetry 1.2.0

tkilias commented on noxfile.py:
```diff
@@ -303,89 +196,148 @@ def stop_db(session: Session) -> None:
+
+
+@nox.session(python=False, name="build-docs")
+def build(session: Session) -> None:
```
build-docs, build ja not specific enough and could be confused with building then project

tkilias commented on noxfile.py:
```diff
@@ -303,89 +196,148 @@ def stop_db(session: Session) -> None:
+
+
+@nox.session(python=False, name="clean-docs")
+def clean(session: Session) -> None:
```
clean_docs because clean could be confused

tkilias commented on noxfile.py:
```diff
@@ -423,29 +375,119 @@ def create_parser() -> ArgumentParser:
+
+@nox.session(name="check-links", python=False)
+def check_links(session: Session) -> None:
+    """Checks weather or not all links in the documentation can be accessed"""
```
I thought sphinx takes care of this

tkilias commented on pyproject.toml:
```diff
@@ -87,8 +87,7 @@ turbodbc = ["turbodbc"]
 [tool.pytest.ini_options]
-addopts= "--tb native -v -r fxX"
-python_files= "test/*test_*.py"
+addopts= "--tb native -v -r fxX --log-debug=sqlalchemy.engine --log-debug=sqlalchemy.pool"
```
Do we have a description what these parameters mean?

Nicoretti replied:
This are options documented and provided by the sqla pytest extension. Would you be in favor of removing them from the general settings?

tkilias replied:
I am find with them for this project, I only would add a comment where they are from

Nicoretti replied:
Created a ticket https://github.com/exasol/sqlalchemy-exasol/issues/219

tkilias commented on setup.cfg:
```diff
@@ -1,3 +1,7 @@
+[db]
+exasol-pyodbc=exa+pyodbc://sys:exasol@localhost:8888/TEST?CONNECTIONCALL=en_US.UTF-8&DRIVER=EXAODBC&SSLCertificate=SSL_VERIFY_NONE
```
A very good, you are setting the locale to a defined value

tkilias commented on sqlalchemy_exasol/merge.py:
```diff
@@ -1,3 +1,8 @@
+# Refactoring notes
```
Are these still needed?

Nicoretti replied:
I would mostly ignore this file for now, with the changed version it needs a complete rework. I can roll it back to its original state though until then. What do you think?

tkilias replied:
ah ok, in that case, I think you can keep the refactoring, because it is way better to understand than the original and might make it easier to rewrite it.

tkilias commented on sqlalchemy_exasol/merge.py:
```diff
@@ -11,115 +16,140 @@
+                for column in self.table.c
+                if column not in self._on_columns and column.name in src_columns
+            )
+            [values.update({column: src_columns[column.name]}) for column in columns]
```
List comprehension should not be used , when the list is not assigned. In that case a normal for loop is usually recommended. An alternative in this case might be a dict or set merge in case values is a dict or set.

tkilias commented on sqlalchemy_exasol/merge.py:
```diff
@@ -11,115 +16,140 @@
-        self._merge_insert_values = ValuesBase(self._target_table, values, [])
+            src_columns = self._source_columns()
+            columns = (column for column in self.table.c if column.name in src_columns)
+            [values.update({column: src_columns[column.name]}) for column in columns]
```
Same as above

tkilias commented on sqlalchemy_exasol/requirements.py:
```diff
@@ -198,3 +198,33 @@ def standalone_null_binds_whereclause(self):
         WHERE clause, in situations where it has to be typed.
         """
         return exclusions.closed()
+
```
I guess this file, was already at this path. However, it actually has to do with tests.

Nicoretti replied:
This must be part of the dialect itself (at least according to the [documentation](https://github.com/sqlalchemy/sqlalchemy/blob/main/README.dialects.rst#dialect-layout))

tkilias replied:
ok, I thought so also first, because it is somewhat a description of capabilities, was only surprised by the testing imports

Nicoretti committed:
Bump poetry versions in github workflows

Nicoretti committed:
Add code review feedback
